activation/wire/malfeasance_double_marry.go

```diff
@@ -0,0 +1,229 @@
+package wire
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/spacemeshos/merkle-tree"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+//go:generate scalegen
+
+// ProofDoubleMarry is a proof that two distinct ATXs contain a marriage certificate signed by the same identity.
+//
+// We are proofing the following:
+// 1. The ATXs have different IDs.
+// 2. Both ATXs have a valid signature.
+// 3. Both ATXs contain a marriage certificate created by the same identity.
+// 4. Both marriage certificates have valid signatures.
+//
+// HINT: this only works if the identity that publishes the ATX with the certificates marries itself.
+type ProofDoubleMarry struct {
+	Proofs [2]MarryProof
+}
+
+var _ Proof = &ProofDoubleMarry{}
+
+func NewDoubleMarryProof(atx1, atx2 *ActivationTxV2, nodeID types.NodeID) (*ProofDoubleMarry, error) {
+	if atx1.ID() == atx2.ID() {
+		return nil, errors.New("ATXs have the same ID")
+	}
+
+	atx1Index := slices.IndexFunc(atx1.Marriages, func(cert MarriageCertificate) bool {
+		return cert.ID == nodeID
+	})
+	if atx1Index == -1 {
+		return nil, errors.New("ATX 1 does not contain a marriage certificate signed by the given node ID")
+	}
+	atx2Index := slices.IndexFunc(atx2.Marriages, func(cert MarriageCertificate) bool {
+		return cert.ID == nodeID
+	})
+	if atx2Index == -1 {
+		return nil, errors.New("ATX 2 does not contain a marriage certificate signed by the given node ID")
+	}
+
+	proof1, err := marriageProof(atx1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create proof for ATX 1: %w", err)
+	}
+
+	proof2, err := marriageProof(atx2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create proof for ATX 2: %w", err)
+	}
+
+	certProof1, err := certificateProof(atx1.Marriages, uint64(atx1Index))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create certificate proof for ATX 1: %w", err)
+	}
+
+	certProof2, err := certificateProof(atx2.Marriages, uint64(atx2Index))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create certificate proof for ATX 2: %w", err)
+	}
+
+	proof := &ProofDoubleMarry{
+		Proofs: [2]MarryProof{
+			{
+				ATXID:  atx1.ID(),
+				NodeID: nodeID,
+
+				MarriageRoot:  types.Hash32(atx1.Marriages.Root()),
+				MarriageProof: proof1,
+
+				CertificateSignature: atx1.Marriages[atx1Index].Signature,
+				CertificateIndex:     uint64(atx1Index),
+				CertificateProof:     certProof1,
+
+				SmesherID: atx1.SmesherID,
```

Comment:
It's super easy to confuse variables here. How about creating a dedicated function to create a `MarryProof`?

```go
func createMarryProof(atx *ActivationTxV2, id types.NodeID) (MarryProof, error) {...}

proof1, err := createMarryProof(atx1, nodeID)
if err != nil {...}
proof2, err := createMarryProof(atx2, nodeID)
if err != nil {...}
proof := &ProofDoubleMarry{Proofs: [2]MarryProof{proof1, proof2}}

Reply:
I refactored this section as suggested